src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -209,6 +209,10 @@ public static MethodHandle adaptUpcallForIMR(MethodHandle target, boolean dropRe
 
         if (dropReturn) { // no handling for return value, need to drop it
             target = dropReturn(target);
+        } else {
+            // adjust return type so it matches the inferred type of the effective
+            // function descriptor
+            target = target.asType(target.type().changeReturnType(Addressable.class));
```

Comment:
Note that this is needed to prevent a situation on Windows where the effective function descriptor of `()[<some IMR struct>]` is the same as a plain `(a64)a64` function descriptor. The `target` will currently have the type `(MemoryAddress)MemoryAddress` in the former case, but `(MemoryAddress)Addressable` in the latter case. However, they share an upcall wrapper class because they have the same function descriptor, and this leads to a `WrongMethodTypeException` when invoking one of them.

Reply:
I think you mean in the latter case is `(Addressable)MemoryAddress` ?

Reply:
No, for upcalls it is `(MemoryAddress)Addressable` i.e. the return type is changed.

Reply:
aaah contravariance :-)